pkg/core/storelimit/limit_test.go

```diff
@@ -123,57 +122,49 @@ func TestFeedback(t *testing.T) {
 	s := NewSlidingWindows()
 	re := require.New(t)
 	type SnapshotStats struct {
-		total     int64
-		remaining int64
-		size      int64
-		start     int64
+		expectCost int64
+		remaining  int64
+		size       int64
+		start      int64
 	}
 	// region size is 10GB, snapshot write limit is 100MB/s and the snapshot concurrency is 3.
 	// the best strategy is that the tikv executing queue equals the wait.
 	const regionSize, limit, wait = int64(10000), int64(100), int64(4)
-	var iter atomic.Int32
-	iter.Store(100)
-	ops := make(chan int64, 10)
+	iter := 100
 	ctx, cancel := context.WithCancel(context.Background())
-
-	// generate the operator
-	go func() {
-		for {
-			if s.Available(regionSize, SendSnapshot, constant.Low) && iter.Load() > 0 {
-				iter.Add(-1)
-				size := regionSize - rand.Int63n(regionSize/10)
-				s.Take(size, SendSnapshot, constant.Low)
-				ops <- size
-			}
-			if iter.Load() == 0 {
-				cancel()
-				return
-			}
-		}
-	}()
+	defer cancel()
 
 	// receive the operator
 	queue := list.List{}
 	interval := time.Microsecond * 100
 	ticker := time.NewTicker(interval)
 	defer ticker.Stop()
 
+	// create one region operator
+	generateOp := func(tick int64) {
+		if s.Available(regionSize, SendSnapshot, constant.Low) && iter > 0 {
+			iter--
+			size := regionSize - rand.Int63n(regionSize/10)
+			stats := &SnapshotStats{
+				expectCost: size / limit,
+				remaining:  size,
+				size:       size,
+				start:      tick,
+			}
+			s.Take(size, SendSnapshot, constant.Low)
+			queue.PushBack(stats)
+		}
+	}
+
 	// tick is the time that the snapshot has been executed.
 	tick := int64(0)
 	for {
 		select {
-		case op := <-ops:
-			stats := &SnapshotStats{
-				total:     op / limit,
-				remaining: op,
-				size:      op,
-				start:     tick,
-			}
-			queue.PushBack(stats)
 		case <-ctx.Done():
```

Comment:
Do we still need `context` here after removing goroutine? I think we can remove it.